From bda8d93ba95f34aa91ae8578fe7d3454f3324f6a Mon Sep 17 00:00:00 2001
From: Peter Dillinger <peterd@fb.com>
Date: Fri, 10 Sep 2021 09:59:05 -0700
Subject: [PATCH] Fix and detect headers with missing dependencies (#8893)

Summary:
It's always annoying to find a header does not include its own
dependencies and only works when included after other includes. This
change adds `make check-headers` which validates that each header can
be included at the top of a file. Some headers are excluded e.g. because
of platform or external dependencies.

rocksdb_namespace.h had to be re-worked slightly to enable checking for
failure to include it. (ROCKSDB_NAMESPACE is a valid namespace name.)

Fixes mostly involve adding and cleaning up #includes, but for
FileTraceWriter, a constructor was out-of-lined to make a forward
declaration sufficient.

This check is not currently run with `make check` but is added to
CircleCI build-linux-unity since that one is already relatively fast.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/8893

Test Plan: existing tests and resolving issues detected by new check

Reviewed By: mrambacher

Differential Revision: D30823300

Pulled By: pdillinger

fbshipit-source-id: 9fff223944994c83c105e2e6496d24845dc8e572
---
 .circleci/config.yml                          |  7 ++---
 Makefile                                      | 27 +++++++++++++++++--
 cache/cache_entry_stats.h                     |  1 +
 db/compaction/compaction_iteration_stats.h    |  2 ++
 db/job_context.h                              |  3 ++-
 db/pre_release_callback.h                     |  3 +--
 db/read_callback.h                            |  1 +
 db/snapshot_impl.h                            |  1 +
 db_stress_tool/db_stress_compaction_filter.h  |  2 ++
 db_stress_tool/db_stress_listener.h           |  3 +++
 .../db_stress_table_properties_collector.h    |  1 +
 include/rocksdb/functor_wrapper.h             |  1 +
 include/rocksdb/rocksdb_namespace.h           |  6 +++++
 include/rocksdb/thread_status.h               |  4 ++-
 include/rocksdb/universal_compaction.h        |  4 ++-
 .../utilities/ldb_cmd_execute_result.h        |  4 +++
 memory/memory_usage.h                         |  3 +++
 table/block_based/block_type.h                |  2 ++
 table/table_properties_internal.h             |  2 +-
 table/table_reader_caller.h                   |  2 ++
 util/cast_util.h                              |  3 +++
 util/channel.h                                |  2 ++
 util/crc32c_ppc.h                             |  3 +++
 util/defer.h                                  |  2 ++
 util/duplicate_detector.h                     |  3 +--
 util/fastrange.h                              |  2 ++
 util/math.h                                   |  2 ++
 util/set_comparator.h                         |  2 ++
 util/work_queue.h                             |  2 ++
 util/xxph3.h                                  |  1 +
 utilities/blob_db/blob_db_gc_stats.h          |  4 +++
 utilities/cassandra/cassandra_options.h       |  5 +++-
 utilities/cassandra/serialize.h               |  5 ++++
 utilities/trace/file_trace_reader_writer.cc   |  4 +++
 utilities/trace/file_trace_reader_writer.h    |  3 +--
 35 files changed, 106 insertions(+), 16 deletions(-)

diff --git a/storage/rocksdb/rocksdb/db/compaction/compaction_iteration_stats.h b/storage/rocksdb/rocksdb/db/compaction/compaction_iteration_stats.h
index 894aea1d96f..910c4469a69 100644
--- a/storage/rocksdb/rocksdb/db/compaction/compaction_iteration_stats.h
+++ b/storage/rocksdb/rocksdb/db/compaction/compaction_iteration_stats.h
@@ -5,6 +5,8 @@
 
 #pragma once
 
+#include <cstdint>
+
 #include "rocksdb/rocksdb_namespace.h"
 
 struct CompactionIterationStats {
diff --git a/storage/rocksdb/rocksdb/db/job_context.h b/storage/rocksdb/rocksdb/db/job_context.h
index 4b0a8edfc02..8e191e31c4d 100644
--- a/storage/rocksdb/rocksdb/db/job_context.h
+++ b/storage/rocksdb/rocksdb/db/job_context.h
@@ -12,8 +12,9 @@
 #include <string>
 #include <vector>
 
-#include "db/log_writer.h"
 #include "db/column_family.h"
+#include "db/log_writer.h"
+#include "db/version_set.h"
 
 namespace ROCKSDB_NAMESPACE {
 
diff --git a/storage/rocksdb/rocksdb/db/pre_release_callback.h b/storage/rocksdb/rocksdb/db/pre_release_callback.h
index b74be953756..6b903948779 100644
--- a/storage/rocksdb/rocksdb/db/pre_release_callback.h
+++ b/storage/rocksdb/rocksdb/db/pre_release_callback.h
@@ -6,11 +6,10 @@
 #pragma once
 
 #include "rocksdb/status.h"
+#include "rocksdb/types.h"
 
 namespace ROCKSDB_NAMESPACE {
 
-class DB;
-
 class PreReleaseCallback {
  public:
   virtual ~PreReleaseCallback() {}
diff --git a/storage/rocksdb/rocksdb/db/read_callback.h b/storage/rocksdb/rocksdb/db/read_callback.h
index 8989c755abf..c042352db0c 100644
--- a/storage/rocksdb/rocksdb/db/read_callback.h
+++ b/storage/rocksdb/rocksdb/db/read_callback.h
@@ -5,6 +5,7 @@
 
 #pragma once
 
+#include "db/dbformat.h"
 #include "rocksdb/types.h"
 
 namespace ROCKSDB_NAMESPACE {
diff --git a/storage/rocksdb/rocksdb/db/snapshot_impl.h b/storage/rocksdb/rocksdb/db/snapshot_impl.h
index bfa44e3f53b..9961bdd584f 100644
--- a/storage/rocksdb/rocksdb/db/snapshot_impl.h
+++ b/storage/rocksdb/rocksdb/db/snapshot_impl.h
@@ -10,6 +10,7 @@
 #pragma once
 #include <vector>
 
+#include "db/dbformat.h"
 #include "rocksdb/db.h"
 
 namespace ROCKSDB_NAMESPACE {
diff --git a/storage/rocksdb/rocksdb/db_stress_tool/db_stress_listener.h b/storage/rocksdb/rocksdb/db_stress_tool/db_stress_listener.h
index 6985a77420d..7fbab57e711 100644
--- a/storage/rocksdb/rocksdb/db_stress_tool/db_stress_listener.h
+++ b/storage/rocksdb/rocksdb/db_stress_tool/db_stress_listener.h
@@ -3,11 +3,14 @@
 //  COPYING file in the root directory) and Apache 2.0 License
 //  (found in the LICENSE.Apache file in the root directory).
 
+#include "file/filename.h"
 #ifdef GFLAGS
 #pragma once
 
+#include "rocksdb/db.h"
 #include "rocksdb/listener.h"
 #include "util/gflags_compat.h"
+#include "util/random.h"
 
 DECLARE_int32(compact_files_one_in);
 
diff --git a/storage/rocksdb/rocksdb/include/rocksdb/thread_status.h b/storage/rocksdb/rocksdb/include/rocksdb/thread_status.h
index 6b2f5c885ae..bb2de6c7b6b 100644
--- a/storage/rocksdb/rocksdb/include/rocksdb/thread_status.h
+++ b/storage/rocksdb/rocksdb/include/rocksdb/thread_status.h
@@ -13,13 +13,15 @@
 
 #pragma once
 
-#include <stdint.h>
 #include <cstddef>
+#include <cstdint>
 #include <map>
 #include <string>
 #include <utility>
 #include <vector>
 
+#include "rocksdb/rocksdb_namespace.h"
+
 #if !defined(ROCKSDB_LITE) && !defined(NROCKSDB_THREAD_STATUS) && \
     defined(ROCKSDB_SUPPORT_THREAD_LOCAL)
 #define ROCKSDB_USING_THREAD_STATUS
diff --git a/storage/rocksdb/rocksdb/include/rocksdb/universal_compaction.h b/storage/rocksdb/rocksdb/include/rocksdb/universal_compaction.h
index f4df5c0009c..9a09d8f4e71 100644
--- a/storage/rocksdb/rocksdb/include/rocksdb/universal_compaction.h
+++ b/storage/rocksdb/rocksdb/include/rocksdb/universal_compaction.h
@@ -5,10 +5,12 @@
 
 #pragma once
 
-#include <stdint.h>
 #include <climits>
+#include <cstdint>
 #include <vector>
 
+#include "rocksdb/rocksdb_namespace.h"
+
 namespace ROCKSDB_NAMESPACE {
 
 //
diff --git a/storage/rocksdb/rocksdb/include/rocksdb/utilities/ldb_cmd_execute_result.h b/storage/rocksdb/rocksdb/include/rocksdb/utilities/ldb_cmd_execute_result.h
index c837b47f772..57bac334682 100644
--- a/storage/rocksdb/rocksdb/include/rocksdb/utilities/ldb_cmd_execute_result.h
+++ b/storage/rocksdb/rocksdb/include/rocksdb/utilities/ldb_cmd_execute_result.h
@@ -5,6 +5,10 @@
 //
 #pragma once
 
+#include <string>
+
+#include "rocksdb/rocksdb_namespace.h"
+
 #ifdef FAILED
 #undef FAILED
 #endif
diff --git a/storage/rocksdb/rocksdb/memory/memory_usage.h b/storage/rocksdb/rocksdb/memory/memory_usage.h
index 9061d45eb44..7f1de8c9772 100644
--- a/storage/rocksdb/rocksdb/memory/memory_usage.h
+++ b/storage/rocksdb/rocksdb/memory/memory_usage.h
@@ -5,8 +5,11 @@
 
 #pragma once
 
+#include <cstddef>
 #include <unordered_map>
 
+#include "rocksdb/rocksdb_namespace.h"
+
 namespace ROCKSDB_NAMESPACE {
 
 // Helper methods to estimate memroy usage by std containers.
diff --git a/storage/rocksdb/rocksdb/table/block_based/block_type.h b/storage/rocksdb/rocksdb/table/block_based/block_type.h
index b2a913746e8..0ca899f2016 100644
--- a/storage/rocksdb/rocksdb/table/block_based/block_type.h
+++ b/storage/rocksdb/rocksdb/table/block_based/block_type.h
@@ -7,6 +7,8 @@
 
 #include <cstdint>
 
+#include "rocksdb/rocksdb_namespace.h"
+
 namespace ROCKSDB_NAMESPACE {
 
 // Represents the types of blocks used in the block based table format.
diff --git a/storage/rocksdb/rocksdb/table/table_properties_internal.h b/storage/rocksdb/rocksdb/table/table_properties_internal.h
index a7a92e3e1fc..1711924349a 100644
--- a/storage/rocksdb/rocksdb/table/table_properties_internal.h
+++ b/storage/rocksdb/rocksdb/table/table_properties_internal.h
@@ -6,7 +6,7 @@
 #pragma once
 
 #include "rocksdb/status.h"
-#include "rocksdb/iterator.h"
+#include "table/internal_iterator.h"
 
 namespace ROCKSDB_NAMESPACE {
 
diff --git a/storage/rocksdb/rocksdb/table/table_reader_caller.h b/storage/rocksdb/rocksdb/table/table_reader_caller.h
index 7a57b5e987c..10ec08130fb 100644
--- a/storage/rocksdb/rocksdb/table/table_reader_caller.h
+++ b/storage/rocksdb/rocksdb/table/table_reader_caller.h
@@ -5,6 +5,8 @@
 
 #pragma once
 
+#include "rocksdb/rocksdb_namespace.h"
+
 namespace ROCKSDB_NAMESPACE {
 // A list of callers for a table reader. It is used to trace the caller that
 // accesses on a block. This is only used for block cache tracing and analysis.
diff --git a/storage/rocksdb/rocksdb/util/channel.h b/storage/rocksdb/rocksdb/util/channel.h
index 705fa2d28f3..19b95629771 100644
--- a/storage/rocksdb/rocksdb/util/channel.h
+++ b/storage/rocksdb/rocksdb/util/channel.h
@@ -10,6 +10,8 @@
 #include <queue>
 #include <utility>
 
+#include "rocksdb/rocksdb_namespace.h"
+
 namespace ROCKSDB_NAMESPACE {
 
 template <class T>
diff --git a/storage/rocksdb/rocksdb/util/crc32c_ppc.h b/storage/rocksdb/rocksdb/util/crc32c_ppc.h
index 15ed6fae5e0..f0b0b66d5db 100644
--- a/storage/rocksdb/rocksdb/util/crc32c_ppc.h
+++ b/storage/rocksdb/rocksdb/util/crc32c_ppc.h
@@ -7,6 +7,9 @@
 
 #pragma once
 
+#include <cstddef>
+#include <cstdint>
+
 #ifdef __cplusplus
 extern "C" {
 #endif
diff --git a/storage/rocksdb/rocksdb/util/defer.h b/storage/rocksdb/rocksdb/util/defer.h
index a781931733d..33c0243e650 100644
--- a/storage/rocksdb/rocksdb/util/defer.h
+++ b/storage/rocksdb/rocksdb/util/defer.h
@@ -7,6 +7,8 @@
 
 #include <functional>
 
+#include "rocksdb/rocksdb_namespace.h"
+
 namespace ROCKSDB_NAMESPACE {
 
 // Defers the execution of the provided function until the Defer
diff --git a/storage/rocksdb/rocksdb/util/set_comparator.h b/storage/rocksdb/rocksdb/util/set_comparator.h
index 9b5cfc1dc5f..e0e64436ade 100644
--- a/storage/rocksdb/rocksdb/util/set_comparator.h
+++ b/storage/rocksdb/rocksdb/util/set_comparator.h
@@ -5,6 +5,8 @@
 
 #pragma once
 
+#include "rocksdb/comparator.h"
+
 namespace ROCKSDB_NAMESPACE {
 // A comparator to be used in std::set
 struct SetComparator {
diff --git a/storage/rocksdb/rocksdb/utilities/blob_db/blob_db_gc_stats.h b/storage/rocksdb/rocksdb/utilities/blob_db/blob_db_gc_stats.h
index 1e6e4a25d74..fea6b003289 100644
--- a/storage/rocksdb/rocksdb/utilities/blob_db/blob_db_gc_stats.h
+++ b/storage/rocksdb/rocksdb/utilities/blob_db/blob_db_gc_stats.h
@@ -5,6 +5,10 @@
 //
 #pragma once
 
+#include <cstdint>
+
+#include "rocksdb/rocksdb_namespace.h"
+
 #ifndef ROCKSDB_LITE
 
 namespace ROCKSDB_NAMESPACE {
diff --git a/storage/rocksdb/rocksdb/utilities/cassandra/serialize.h b/storage/rocksdb/rocksdb/utilities/cassandra/serialize.h
index cd980ade0f6..8f50a02dd97 100644
--- a/storage/rocksdb/rocksdb/utilities/cassandra/serialize.h
+++ b/storage/rocksdb/rocksdb/utilities/cassandra/serialize.h
@@ -10,6 +10,11 @@
 
 #pragma once
 
+#include <cstdint>
+#include <string>
+
+#include "rocksdb/rocksdb_namespace.h"
+
 namespace ROCKSDB_NAMESPACE {
 namespace cassandra {
 namespace {
diff --git a/storage/rocksdb/rocksdb/table/block_based/data_block_hash_index.h b/storage/rocksdb/rocksdb/table/block_based/data_block_hash_index.h
--- a/storage/rocksdb/rocksdb/table/block_based/data_block_hash_index.h	2023-02-06 19:55:30.000000000 +0000
+++ b/storage/rocksdb/rocksdb/table/block_based/data_block_hash_index.h	2023-04-14 22:26:58.371446661 +0000
@@ -5,6 +5,7 @@
 
 #pragma once
 
+#include <cstdint>
 #include <string>
 #include <vector>
 
diff --git a/storage/rocksdb/rocksdb/util/string_util.h b/storage/rocksdb/rocksdb/util/string_util.h
--- a/storage/rocksdb/rocksdb/util/string_util.h	2023-02-06 19:55:30.000000000 +0000
+++ b/storage/rocksdb/rocksdb/util/string_util.h	2023-04-14 22:33:07.714153055 +0000
@@ -6,6 +6,7 @@
 
 #pragma once
 
+#include <cstdint>
 #include <sstream>
 #include <string>
 #include <unordered_map>
--- a/storage/rocksdb/rocksdb/utilities/checkpoint/checkpoint_impl.h	2023-02-06 19:55:30.000000000 +0000
+++ b/storage/rocksdb/rocksdb/utilities/checkpoint/checkpoint_impl.h	2023-04-14 22:38:35.569394128 +0000
@@ -8,6 +8,7 @@
 
 #include "rocksdb/utilities/checkpoint.h"
 
+#include <cstdint>
 #include <string>
 #include "file/filename.h"
 #include "rocksdb/db.h"
--- a/storage/rocksdb/rocksdb/include/rocksdb/utilities/checkpoint.h	2023-02-06 19:55:30.000000000 +0000
+++ b/storage/rocksdb/rocksdb/include/rocksdb/utilities/checkpoint.h	2023-02-06 19:55:30.000000000 +0000
@@ -8,6 +8,7 @@
 #pragma once
 #ifndef ROCKSDB_LITE
 
+#include <cstdint>
 #include <string>
 #include <vector>
 #include "rocksdb/status.h"
